File: 01_Search/BinarySearch/BinarySearch.cpp
#include <iostream>
#include <vector>
#include <time.h>
#include<algorithm>
using namespace std;

/*
		creating an array of size n
		initializing the array with random values
*/
vector<int> initializeArrayWithRandomNumbers(int n)
{
    srand (time(NULL));
    vector<int> vec;
    for(int i = 0; i<n; ++i)
    {
        vec.push_back(rand()%n+1);
    }

    return vec;
}//initializeArrayWithRandomNumbers

void printVector(vector<int>vec)
{
    for(int i =0; i<vec.size(); ++i)
    {
        cout<<vec.at(i)<<"\t";
    }

    cout<<endl;
}//printVector

int binarySearch(vector<int>arr, int target, int low, int high)
{
    if(low>high)
        return -1;
    int mid = (low + high) / 2;
    if(target == arr.at(mid))
    {
        return mid;
    }

    if(target < arr.at(mid))
    {
        binarySearch(arr, target, low, mid);
    }

    if(target > arr.at(mid))
    {
        binarySearch(arr, target, mid+1, high);
    } 
}//binarySearch

int main()
{
    cout<<"The array that needs to be searched through\n";
    int size = 10;
    vector<int>vec = initializeArrayWithRandomNumbers(size);
    printVector(vec);

    cout<<"**********Binary Search**********\n";
    srand (time(NULL));
    int tar = (rand()%size+1);
    cout<<"Value to be searched: "<<tar<<endl;
    sort(vec.begin(), vec.end());
    cout<<"Sorted Array\n";
    printVector(vec);
    int ans = binarySearch(vec, tar, 0, vec.size());
    cout<<"Value present on Index: "<<ans<<endl;
    
    return 0;
}//main

File: 01_Search/BinarySearch/BinarySearch.py
import random
 
"""
 creating an array of size n
 initializing the array with random values
"""
def initializeArrayWithRandomNumbers(n):
	myList = []
	for i in range(n):
		myList.append(random.randint(0,n))
	return myList
	
def binarySearch(myList, target, low, high):
	if low>high:
		return -1
	mid = int((low + high) / 2)
	if target == myList[mid]:
		return mid
	if target < myList[mid]:
		return binarySearch(myList, target, low, mid)
	if target > myList[mid]:
		return binarySearch(myList, target, mid+1, high)
	
size = 10
print("Array that needs to be searched through")
myList = initializeArrayWithRandomNumbers(size)
print(myList)

print("**********Binary Search**********")
myList.sort()
tar = random.randint(0,size)
print ("Sorted Array")
print(myList)
print("Value to be found: ", tar)
myPos = binarySearch(myList, tar, 0, len(myList)-1)
print("Value present on Index: ", myPos)